docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+the same generation validation requirement.
+
+### Pageserver attach/startup changes
+
+#### Pageserver node startup
+
+- The pageserver must obtain a node generation number by some means (see Control Plane Changes below) before
+  doing any remote writes.
+- The pageserver _may_ also do some synchronization of its attachments to see which are still
+  valid, see "Synchronizing attachments on pageserver startup" in the optimizations section.
+
+#### Attachment
+
+Calls to `/v1/tenant/{tenant_id}/attach` are augmented with generation details:
+
+- Required: provide an attachment generation number
+- Nice to have: provide the attachment generation, node id, and node generation of
+  where the tenant was previously attached, to help this node calculate where
+  to find the latest index_part file.
+
+The pageserver must persist this attachment generation number before starting
+the `Tenant` tasks. We will create a new per-tenant metadata file that records
+the attachment generation number, and will also be used in future work for
+implementing fast migration, where the tenant has more states than simple
+attached & detached.
+
+#### Reconciliation
+
+On attachment, the `Tenant` may have some local disk content from previous
+activity involving this tenant (e.g. it used to be attached here), and this
+needs reconciling with remote storage. Note that "reconcile" in this context is
+used generically, and does not refer specifically to the logic currently
+in the `reconcile_with_remote` function.
+
+Because pageservers have access to the WAL, they don't strictly have to
+use the latest remote storage contents at all: if the WAL history still contains
+the data required to bring the pageserver's local state up to date with
+the tip of the WAL, the pageserver could recover from the WAL and then
+overwrite whatever is in remote storage.
+
+However, it is preferable to recover from remote storage rather than
+to replay WAL, for the following reasons:
+
+- Simplicity: recovering from remote storage gives identical behavior
+  whether attaching a tenant for the first time or re-attaching later.
+- Efficiency: ingesting the WAL has significant cost beyond the I/O
+  of reading it (compaction, image layer generation). We should avoid
+  doing this work twice if there is already remote content available.
+- Reduced load on safekeepers, which are a less scalable resource than S3.
+
+#### Finding the remote indices for timelines
+
+Because index files are now suffixed with generation numbers, the pageserver
+cannot always GET the remote index in one request, because it can't always
+know a-priori what the latest remote index is.
+
+In the general case and as a fallback, the pageserver may list all the index*part.json
+files for a timeline, sort them by generation suffix, and pick the highest that is <=
+the suffix for its current generation numbers. The tenant should never load an index with
+an attachment generation \_newer* than its own: tenants
```

Comment:
Some weird escaping going on?

Plus, let's just use "its suffix" instead of "its current generation numbers".

```suggestion
In the general case and as a fallback, the pageserver may list all the `index_part.json``
files for a timeline, sort them by generation suffix, and pick the highest that is `<=`
its suffix for this attachment. The tenant should never load an index with
a suffix newer than its own: tenants
```

Reply:
Also, I first wanted to suggest to use `<` instead of `<=`, only to realize then that in the future we may want to implement support for reloading a tenant during a process lifetime. In fact we already have the `/ignore` and `/load` endpoints that do that.

Which makes me wonder: to guard against bugs with `/ignore` (e.g., cancellation not working correctly), wouldn't we want similar split-brain protection as well? (i.e., e.g., get a new attachment generation number from the control plane?)

Reply:
Tweaked by hand 666164ab (it's easier to phrase this now that there is only one generation in play).

>Which makes me wonder: to guard against bugs with /ignore (e.g., cancellation not working correctly), wouldn't we want similar split-brain protection as well? (i.e., e.g., get a new attachment generation number from the control plane?)

Hmm, `/load` and `/ignore` are kinda weird.  When we do multiple locations per tenant, they would just become "secondary locations" I think, i.e. files on disk but don't act like an attachment.  We'll get into it when implementing I think -- maybe `/load` will need to take a generation number like attach does, as it's essentially just an "attach but with existing files", and ignore is just a "like detach but don't delete the files".